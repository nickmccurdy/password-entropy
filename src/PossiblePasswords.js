import { faList, faRandom } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import {
  Card,
  CardBody,
  FormGroup,
  FormText,
  Label,
  Progress,
} from 'reactstrap'

export default function PossiblePasswords({
  possiblePasswords,
  approximate,
  entropyBits,
  entropyTip: { strength },
}) {
  return (
    <Card>
      <CardBody>
        <FormGroup>
          <Label>
            <FontAwesomeIcon icon={faList} /> Entropy
          </Label>

          <FormText>
            <Progress max={128} value={entropyBits}>
              {entropyBits.toFixed(2)} bits ({strength})
            </Progress>
          </FormText>
        </FormGroup>

        <FormGroup>
          <Label>
            <FontAwesomeIcon icon={faRandom} /> Possible Passwords
          </Label>

          <FormText>
            {approximate && '~ '} {possiblePasswords}
          </FormText>
        </FormGroup>
      </CardBody>
    </Card>
  )
}

PossiblePasswords.propTypes = {
  approximate: PropTypes.bool.isRequired,
  entropyBits: PropTypes.number.isRequired,
  entropyTip: PropTypes.shape({
    minimum: PropTypes.number.isRequired,
    strength: PropTypes.string.isRequired,
    style: PropTypes.string.isRequired,
  }).isRequired,
  possiblePasswords: PropTypes.number.isRequired,
}
