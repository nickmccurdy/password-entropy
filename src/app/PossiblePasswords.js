import { Card, CardText, LinearProgress } from 'material-ui'
import { colors } from 'material-ui/styles'
import PropTypes from 'prop-types'
import { always, filter, findLast, identity, keys, map, pipe, sum } from 'ramda'
import React from 'react'
import Icon from 'react-fa'
import { connect } from 'react-redux'

import { dicewareWords } from '../options/DicewareOptions'
import { toggles } from '../options/GenericOptions'

const entropyTips = [
  {
    color: colors.red500,
    minimum: 0,
    strength: 'Very Weak'
  },
  {
    color: colors.yellow500,
    minimum: 32,
    strength: 'Weak'
  },
  {
    color: colors.blue500,
    minimum: 64,
    strength: 'Strong'
  },
  {
    color: colors.green500,
    minimum: 128,
    strength: 'Very Strong'
  }
]

export function PossiblePasswords({
  possiblePasswords, approximate, entropyBits, entropyTip: { color, strength }
}) {
  return (
    <Card>
      <CardText>
        <h3><Icon name="list" /> Entropy</h3>
        {entropyBits.toFixed(2)} bits ({strength})
        <LinearProgress color={color} max={128} mode="determinate" value={entropyBits} />

        <h3><Icon name="random" /> Possible Passwords</h3>
        {approximate && '~ '}{possiblePasswords.toLocaleString()}
      </CardText>
    </Card>
  )
}

PossiblePasswords.propTypes = {
  approximate: PropTypes.bool.isRequired,
  entropyBits: PropTypes.number.isRequired,
  entropyTip: PropTypes.shape({
    color: PropTypes.string.isRequired,
    minimum: PropTypes.number.isRequired,
    strength: PropTypes.string.isRequired
  }).isRequired,
  possiblePasswords: PropTypes.number.isRequired
}

const possibleItems = {
  generic: pipe(
    filter(identity),
    keys,
    map(key => toggles[key].possibleItems),
    sum
  ),
  diceware: always(dicewareWords),
  custom: identity
}

export function mapStateToProps({ length, options, optionsKey }) {
  const possiblePasswords = possibleItems[optionsKey](options[optionsKey]) ** length
  const entropyBits = Math.max(0, Math.log2(possiblePasswords))

  return {
    approximate: possiblePasswords > Number.MAX_SAFE_INTEGER,
    entropyBits,
    entropyTip: findLast(tip => entropyBits >= tip.minimum, entropyTips),
    possiblePasswords
  }
}

export default connect(mapStateToProps)(PossiblePasswords)
