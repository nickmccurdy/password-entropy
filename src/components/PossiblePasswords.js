import { faList, faRandom } from '@fortawesome/fontawesome-free-solid'
import Icon from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import React from 'react'
import {
  Card,
  CardBody,
  FormGroup,
  FormText,
  Label,
  Progress
} from 'reactstrap'

export default function PossiblePasswords({
  possiblePasswords,
  approximate,
  entropyBits,
  entropyTip: { strength, style }
}) {
  return (
    <Card>
      <CardBody>
        <FormGroup>
          <Label>
            <Icon icon={faList} /> Entropy
          </Label>

          <FormText>
            <Progress max={128} value={entropyBits}>
              {entropyBits.toFixed(2)} bits ({strength})
            </Progress>
          </FormText>
        </FormGroup>

        <FormGroup>
          <Label>
            <Icon icon={faRandom} /> PossiblePasswords
          </Label>

          <FormText>
            {approximate && '~ '} {possiblePasswords.toLocaleString()}
          </FormText>
        </FormGroup>
      </CardBody>
    </Card>
  )
}

PossiblePasswords.propTypes = {
  approximate: PropTypes.bool.isRequired,
  entropyBits: PropTypes.number.isRequired,
  entropyTip: PropTypes.shape({
    minimum: PropTypes.number.isRequired,
    strength: PropTypes.string.isRequired,
    style: PropTypes.string.isRequired
  }).isRequired,
  possiblePasswords: PropTypes.number.isRequired
}
