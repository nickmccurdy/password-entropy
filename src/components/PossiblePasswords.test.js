import blue from 'material-ui/colors/blue'
import React from 'react'
import Renderer from 'react-test-renderer/shallow'

import { PossiblePasswords } from './PossiblePasswords'

describe('PossiblePasswords', () => {
  function getOutput(approximate) {
    return new Renderer().render(<PossiblePasswords approximate={approximate} entropyBits={77.54887502163469} entropyTip={{ color: blue, minimum: 64, strength: 'Strong' }} possiblePasswords={2.2107391972073336e+23} />)
  }

  test('approximate', () => {
    expect(getOutput(true)).toMatchSnapshot()
  })

  test('exact', () => {
    expect(getOutput(false)).toMatchSnapshot()
  })
})
