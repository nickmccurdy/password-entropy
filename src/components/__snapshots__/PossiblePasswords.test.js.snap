// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PossiblePasswords approximate 1`] = `
<Panel
  bsClass="panel"
  bsStyle="default"
  defaultExpanded={false}
>
  <FormGroup
    icon="list"
    id="entropy"
    label="Entropy"
  >
    <ProgressBar
      active={false}
      bsClass="progress-bar"
      bsStyle="info"
      isChild={false}
      label="77.55 bits (Strong)"
      max={128}
      min={0}
      now={77.54887502163469}
      srOnly={false}
      striped={false}
    />
  </FormGroup>
  <FormGroup
    icon="random"
    id="possiblePasswords"
    label="PossiblePasswords"
  >
    <FormControlStatic
      bsClass="form-control-static"
      componentClass="p"
    >
      ~ 
      221,073,919,720,733,000,000,000
    </FormControlStatic>
  </FormGroup>
</Panel>
`;

exports[`PossiblePasswords exact 1`] = `
<Panel
  bsClass="panel"
  bsStyle="default"
  defaultExpanded={false}
>
  <FormGroup
    icon="list"
    id="entropy"
    label="Entropy"
  >
    <ProgressBar
      active={false}
      bsClass="progress-bar"
      bsStyle="info"
      isChild={false}
      label="77.55 bits (Strong)"
      max={128}
      min={0}
      now={77.54887502163469}
      srOnly={false}
      striped={false}
    />
  </FormGroup>
  <FormGroup
    icon="random"
    id="possiblePasswords"
    label="PossiblePasswords"
  >
    <FormControlStatic
      bsClass="form-control-static"
      componentClass="p"
    >
      221,073,919,720,733,000,000,000
    </FormControlStatic>
  </FormGroup>
</Panel>
`;
